lucene/core/src/java/org/apache/lucene/geo/Polygon.java

```diff
@@ -96,34 +99,50 @@ public Polygon(double[] polyLats, double[] polyLons, Polygon... holes) {
     this.holes = holes.clone();
 
     // compute bounding box
-    double minLat = polyLats[0];
-    double maxLat = polyLats[0];
-    double minLon = polyLons[0];
-    double maxLon = polyLons[0];
+    double minLat = Double.POSITIVE_INFINITY;
+    double maxLat = Double.NEGATIVE_INFINITY;
+    double minLon = Double.POSITIVE_INFINITY;
+    double maxLon = Double.NEGATIVE_INFINITY;
 
     double windingSum = 0d;
     final int numPts = polyLats.length - 1;
-    for (int i = 1, j = 0; i < numPts; j = i++) {
+    for (int i = 0; i < numPts; i++) {
       minLat = Math.min(polyLats[i], minLat);
       maxLat = Math.max(polyLats[i], maxLat);
       minLon = Math.min(polyLons[i], minLon);
       maxLon = Math.max(polyLons[i], maxLon);
       // compute signed area
-      windingSum += (polyLons[j] - polyLons[numPts])*(polyLats[i] - polyLats[numPts])
-          - (polyLats[j] - polyLats[numPts])*(polyLons[i] - polyLons[numPts]);
+      windingSum += polyLons[i] * polyLats[i + 1] - polyLats[i] * polyLons[i + 1];
+    }
+    if (windingSum == 0) {
+      throw new IllegalArgumentException("Cannot compute the polygon / hole orientation.");
     }
     this.minLat = minLat;
     this.maxLat = maxLat;
     this.minLon = minLon;
     this.maxLon = maxLon;
     this.windingOrder = (windingSum < 0) ? GeoUtils.WindingOrder.CCW : GeoUtils.WindingOrder.CW;
+    double area = Math.abs(windingSum / 2d);
+    for (Polygon hole : holes) {
+      area -= hole.area();
+    }
+    if (area <= 0) {
+      throw new IllegalArgumentException("Polygon has an invalid area (area = " + area + ").");
+    }
+    this.area = area;
+
   }
 
   /** returns the number of vertex points */
   public int numPoints() {
     return polyLats.length;
   }
 
+  /** returns the area of the polygon */
+  public double area() {
```

Comment:
I can imagine some asking what the units are.  And why not getArea?  Maybe this should be marked @lucene.internal or package local even if we don't want to expose it.

Reply:
I rename the method to getArea() and make it protected. I have no use case outside of validation so it is ok to make it package private.

Reply:
I agree w/ @dsmiley re: units.  The calculation is using the [shoelace method](https://en.wikipedia.org/wiki/Shoelace_formula) on vertices in decimal degrees. So the units are deg^2. I remember exploring this a little over a year ago in [LUCENE-8364](https://issues.apache.org/jira/browse/LUCENE-8364) and originally having area because it was "easy". I ended up removing it until we had support for indexing in other projections (e.g., [equal area](https://en.wikipedia.org/wiki/Map_projection#Equal-area)). I think we should wait until [LUCENE-8632](https://issues.apache.org/jira/browse/LUCENE-8632) lands and only provide this area calculation in [XYPolygon](lucene/sandbox/src/java/org/apache/lucene/geo/XYPolygon.java) for regular cartesian geometries in non decimal degrees.

Reply:
I agree we should not provide this info to the user, +1 not to offer the calculation to the user.

The idea behind this change was validation of the polygon validation. The idea of this change came from the fact that I am using this calculation to check that a Polygon tessellation is correct and I realise that:

- Currently we compute the signed area to check the orientation of a polygon, if the area is 0 the polygon is consider CW, should we fail instead?

- I have seen polygons with one hole where the hole and polygon are the same, should we capture that situation and fail?

Reply:
Please add a comment to the code to to link to the "shoelace method" that Nick points at.  I think it's important for non-trivial algorithms to incorporate references where the readers of this code can learn more.  Who knows... maybe some "velcro method" will come along and be superior and then some student looking at this code will point this out to us :-)

Reply:
> * Currently we compute the signed area to check the orientation of a polygon, if the area is 0 the polygon is considered CW, should we fail instead?

+1 to fail in this situation since it is not a valid polygon. 


> * I have seen polygons with one hole where the hole and polygon are the same, should we capture that situation and fail?

+1 to fail in this situation as well.